File: 1285-balance-a-binary-search-tree/1285-balance-a-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func balanceBST(root *TreeNode) *TreeNode {
	if root == nil {
		return nil
	}

	arr := make([]int, 0)
	inorder(&arr, root)

	return constructBST(&arr, 0, len(arr)-1)
}

func constructBST(arr *[]int, l, r int) *TreeNode {
	if l > r {
		return nil
	}

	mid := l + (r-l)/2
	node := &TreeNode{}

	node.Val = (*arr)[mid]

	node.Left = constructBST(arr, l, mid-1)
	node.Right = constructBST(arr, mid+1, r)

	return node
}

func inorder(arr *[]int, node *TreeNode) {
	if node == nil {
		return
	}

	inorder(arr, node.Left)
	*arr = append(*arr, node.Val)
	inorder(arr, node.Right)
}

File: 1285-balance-a-binary-search-tree/1285-balance-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void inorder(TreeNode* node, vector<int>& arr) {
        if (!node)
            return;
        inorder(node->left, arr);
        arr.push_back(node->val);
        inorder(node->right, arr);
    }

    TreeNode* balanceBST(TreeNode* root) {
        vector<int> arr;
        inorder(root, arr);
        return build(0, arr.size() - 1, arr);
    }

    TreeNode* build(int l, int r, vector<int>& arr) {
        if (l > r)
            return NULL;
        int mid = l + (r - l) / 2;
        TreeNode* node = new TreeNode(arr[mid]);
        node->left = build(l, mid - 1, arr);
        node->right = build(mid + 1, r, arr);

        return node;
    }
};